File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/cache/operation/InitializationOperation.kt
/**
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.cache.operation

import ch.nevis.mobile.authentication.sdk.react.cache.state.NoOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.OperationState
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.userinteraction.onerror.DefaultOnErrorImpl
import ch.nevis.mobile.authentication.sdk.react.userinteraction.onsuccess.DefaultOnSuccessImpl
import ch.nevis.mobile.sdk.api.MobileAuthenticationClientError

data class InitializationOperation(
    override var state: OperationState,
    override val onSuccess: (provider: Unit?) -> Unit,
    override val onError: (MobileAuthenticationClientError) -> Unit
) : Operation<Unit>() {
    constructor(operationId: String) : this(
        state = NoOperation,
        onSuccess = { DefaultOnSuccessImpl(operationId, ReactMethodName.INIT_CLIENT).onSuccess() },
        onError = { DefaultOnErrorImpl(operationId, ReactMethodName.INIT_CLIENT).onError(it) }
    )

    override fun cancel() {
        state.cancel()
    }
}

File: ios/Cache/Operation/InitializationOperation.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

class InitializationOperation: Operation {
	// MARK: Properties

	var state: OperationState
	var onError: ErrorHandler
	let onSuccess: () -> ()

	// MARK: Initialization

	init(operationId: String) {
		self.state = NoOperationState()
		self.onSuccess = { OnSuccessImpl(operationId: operationId, method: .initClient).onSuccess() }
		self.onError = { OnErrorImpl(operationId: operationId, method: .initClient).onError($0) }
	}

	// MARK: Operation

	func cancel() {
		state.cancel()
	}
}
